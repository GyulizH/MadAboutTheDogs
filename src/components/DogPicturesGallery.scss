.uploaded-Image-Preview{
  text-align: center;
  margin: 50px 15px;
  height: 400px;
  width: 80%;
  border-left: 1px solid gray;
  border-right: 1px solid gray;
  border-top: 5px solid gray;
  border-bottom: 5px solid gray;
  img{
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
}

.Upload-Img-Label{
  font-size: 20px;
  padding: 20px;
}

.Upload-Img-Wrapper{
  margin: 20px;
}

.DogPicturesGalery-Container{
  display: flex;
}

.DogPictureUpload-Wrapper{
  width: 40%;
  display: inline-block;
  border-right: solid black 1px;
}

.DogPictureBreed{

}
//input[type="file"] {
//  display: none;
//}

.upload-Image-Input {
  //visibility:hidden;
}
.DogPicturesList{
  width: 50%;
  display: inline-block;
  margin-left:10px;
}

.Dog-Picture-Item{
  img{
    height: 200px;
  }
}

.Reset-Img{
  border: 1px solid #ccc;
  display: inline-block;
  padding: 6px 12px;
  cursor: pointer;
}

