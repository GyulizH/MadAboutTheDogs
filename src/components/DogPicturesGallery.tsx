import * as React from 'react';
import './DogPicturesGalery.scss';
import * as mobilenet from '@tensorflow-models/mobilenet';
import * as tf from "@tensorflow/tfjs";




interface State {
  file?: string;
  imgUrl?:string | ArrayBuffer;
  model?:any;
  results?:any;
  dogPicsUrls?:any;
  page?:number;
  loading?:boolean;
  prevY?:number;
  numberOfItems?:number;
};


export default class DogPicturesGalery extends React.Component <{}, State>{
   imageRef: React.RefObject<HTMLImageElement>
   loadingRef: React.RefObject<HTMLDivElement>

   private observer: IntersectionObserver | undefined;

  constructor(props:any) {
    super(props);
    this.state = {
      file: '',
      imgUrl:'',
      model: null,
      results:null,
      dogPicsUrls: [],
      page:0,
      loading:true,
      prevY: 0,
      numberOfItems:10
    }
    this.imageRef = React.createRef()
    this.loadingRef = React.createRef()
    this.resetUploadedImg = this.resetUploadedImg.bind(this)

  }


  async componentDidMount(): Promise <any> {
    tf.getBackend()
    const model = await this.loadModel()
    this.setState({
      model:model
    } as State)
    console.log(tf.getBackend());
    this.observer = new IntersectionObserver(([entry]) => {
       if(entry.isIntersecting){
         this.setState({ loading: true });
         setTimeout(() => {
           this.setState({ numberOfItems: this.state.numberOfItems + 8, loading: false });
         }, 500);
         }
       console.log("LOADINGGG...")
      },
      {
        root: document.getElementsByClassName("main")[0],
        rootMargin: '0px',
        threshold: 1.0
      });

     this.observer.observe(this.loadingRef.current)

  }

  showDogImages (){
    let items = []
    for (let i = 0; i < this.state.numberOfItems; i++) {
      items.push(<div key={i}>
        <img  src={this.state.dogPicsUrls[i]} loading="lazy" />
      </div>);
    }
    return items;
  }
  async identify () {
    const results = await this.state.model.classify(this.imageRef.current);
    this.setState({
      results: results
     } as State, () => this.fetchDogListByBreed())
  }

  resetUploadedImg(){
    this.setState({
       imgUrl: '',
      results: [],
      dogPicsUrls: []
    })
    console.log("helloo")
  }

  fetchDogListByBreed (){
    fetch(`https://dog.ceo/api/breed/hound/images`)
      .then(response => response.json())
      .then(data => {
        let tmpArray = []
        for(let i = 0;i<data.message.length;i++){
          tmpArray.push(data.message[i])
        }
        //use spread operator
        this.setState({
          dogPicsUrls: tmpArray,
         // loading:false
        } as State)
      } )
  }

  async loadModel(){
    return await mobilenet.load();
  }

  handleSubmit(e:any) {
    e.preventDefault();
  }

  handleImageChange(e:any){
    e.preventDefault()
    let reader =  new FileReader()
    let file = e.target.files[0]
    reader.onload = () => {
      this.setState({
        file:file,
        imgUrl:reader.result
      }, async () => {await this.identify()})
    }
    reader.readAsDataURL(file)
  }
  render(){
    const loadingCSS = {
      height: '100px',
      margin:"30px"
    }
    const imgPreviewStyle = {
      padding: '70px',
      fontSize: '20px'
    }
    const loadingTextCSS = {display: this.state.loading ? "block" : "none"}
    let {imgUrl} = this.state
    let imgPreview = null
    if(imgUrl) {
      imgPreview = (<img src={imgUrl as string} ref={this.imageRef}/>)
    }else{
      imgPreview = (<div style={imgPreviewStyle}>Please select an Image for Preview</div>)
    }
    return(
      <div className="DogPicturesGalery-Container">
      <div className="DogPictureUpload-Wrapper">
        <div className="Upload-Img-Wrapper">
          <form onSubmit={(e)=>this.handleSubmit(e)}>
            <label htmlFor="files" className="Upload-Img-Label">Select Image</label>
            <input className="upload-Image-Input"
                   type="file"
                   onChange={(e)=>this.handleImageChange(e)}
            />
          </form>
        </div>
        <div className="uploaded-Image-Preview">
          {imgPreview}
        </div>
        <div className="DogPictureBreed">
          {this.state.results && <span>Results</span>}
          <ul>
            {this.state.results?.map((result:any) => {
              return(<li>
                {result.className}
              </li>)
            })}
          </ul>
        </div>
        <button className="Reset-Img" onClick={this.resetUploadedImg}>RESET</button>
      </div>
      <div className="DogPicturesList">
        <p>DOG PICTURES LIST</p>
        <div className="Dog-Picture-Item">
          <div>
            {this.showDogImages()}
          </div>
        </div>
        <div
          ref={this.loadingRef}
          style={loadingCSS}
        >
          <span style={loadingTextCSS}>
            Loading Images...
          </span>
        </div>
      </div>
      </div>
    )
  }
}

