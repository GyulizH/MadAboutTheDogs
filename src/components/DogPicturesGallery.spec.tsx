import React from 'react'
import renderer from 'react-test-renderer';
import DogPicturesGallery from './DogPicturesGallery';

describe('DogPicturesGallery', ():void => {
  it('should render correctly', ():void => {
     /* Test the state after async functions, write a function to be called  */
     /* After the async calls finished inside dogpicturesgallery */
    expect.assertions(1);
    const wrapper = renderer.create(<DogPicturesGallery/>)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })
})
