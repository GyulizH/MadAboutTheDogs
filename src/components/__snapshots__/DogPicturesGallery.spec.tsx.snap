// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DogPicturesGallery should render correctly 1`] = `
<div
  className="DogPicturesGalery-Container"
>
  <div
    className="DogPictureUpload-Wrapper"
  >
    <div
      className="Upload-Img-Wrapper"
    >
      <form
        onSubmit={[Function]}
      >
        <label
          className="Upload-Img-Label"
          htmlFor="files"
        >
          Select Image
        </label>
        <input
          className="upload-Image-Input"
          onChange={[Function]}
          type="file"
        />
      </form>
    </div>
    <div
      className="uploaded-Image-Preview"
    >
      <div
        style={
          Object {
            "fontSize": "20px",
            "padding": "70px",
          }
        }
      >
        Please select an Image for Preview
      </div>
    </div>
    <div
      className="DogPictureBreed"
    >
      <ul />
    </div>
    <button
      className="Reset-Img"
      onClick={[Function]}
    >
      RESET
    </button>
  </div>
  <div
    className="DogPicturesList"
  >
    <p>
      DOG PICTURES LIST
    </p>
    <div
      className="Dog-Picture-Item"
    >
      <div>
        <div>
          <img
            loading="lazy"
          />
        </div>
        <div>
          <img
            loading="lazy"
          />
        </div>
        <div>
          <img
            loading="lazy"
          />
        </div>
        <div>
          <img
            loading="lazy"
          />
        </div>
        <div>
          <img
            loading="lazy"
          />
        </div>
        <div>
          <img
            loading="lazy"
          />
        </div>
        <div>
          <img
            loading="lazy"
          />
        </div>
        <div>
          <img
            loading="lazy"
          />
        </div>
        <div>
          <img
            loading="lazy"
          />
        </div>
        <div>
          <img
            loading="lazy"
          />
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "height": "100px",
          "margin": "30px",
        }
      }
    >
      <span
        style={
          Object {
            "display": "block",
          }
        }
      >
        Loading Images...
      </span>
    </div>
  </div>
</div>
`;
